### ISSUE TITLE
Training model with batch size 40

### ISSUE THREAD

**Issue by rahilwazir** [torralba-lab/im2recipe-Pytorch#14]
I've executed the code `python train.py --batch_size 40` with the batch size 40 (since my GPU was failing on 160 and 80).

![image](https://user-images.githubusercontent.com/1818941/49698937-6af38d00-fbec-11e8-9838-7cbb7a2783d7.png)

I estimated 12 days since I have lowered the batch size about 3x time and again it's like 14 days now. Can you please advise, when will it be done?

**Comment by nhynes** [comment]
In an unexpected turn of events, I should be receiving the code soon, actually! I'll keep you posted.

**Comment by rahilwazir** [comment]
@nhynes 

OK, I will check it out. Meanwhile can you please tell me how do I know if the model is fully trained?
I have spinned up another VM with 4x12GB Tesla K80 cards and it's still taking long time to complete with `... -batch_size=240`

![image](https://user-images.githubusercontent.com/1818941/50794545-3c50fb00-12ed-11e9-8b8c-acd575d413dc.png)

Please advise.

**Comment by nhynes** [comment]
It looks like your optimization is stuck. Are you sure you're running the model on the Recipe1M dataset?

**Comment by nhynes** [comment]
What happens if you try training using the recommended parameters
```
python train.py 
--img_path /path/to/images/ 
--data_path /path/to/lmdbs/ 
--ingrW2V /path/to/w2v/vocab.bin
--snapshots snapshots/
--valfreq 10
```

Also, what's your goal? Can you just use the pretrained model?

**Comment by rahilwazir** [comment]
OK, I have restarted the training with the following command now

```
$ python train.py --ingrW2V data/vocab.bin --valfreq 10 --snapshots snapshots/ --data_path data/ --img_path data/recipe1M/
```

> Also, what's your goal?

I just want a trained model which I would like to test against the receipes.

> Can you just use the pretrained model?

I can, but I tried previously with `test.py` script, it triggered some error. I have ran it again. Let's see what does it output this time.

**Comment by nhynes** [comment]
what was the error that was triggered? training a copy of the trained model is going to yield the same error.

**Comment by rahilwazir** [comment]
Previously it spits the following, when I had only single 12GB GPU

> RuntimeError: CUDA error: out of memory

I have not tried it with the latest VM specs, as I said, I'll do that if the process halts again.

**Comment by rahilwazir** [comment]
It's been almost a week now and the model is still training. Looks like it's NOT stuck anymore but it is still taking long time to complete. What do you suggest here?

![image](https://user-images.githubusercontent.com/1818941/51086248-3e401180-1766-11e9-87e9-66d9551f7179.png)

**Comment by felicityji** [comment]
@rahilwazir   hi, Did you run into the same error during training the model?
![screenshot from 2019-01-18 13-38-26](https://user-images.githubusercontent.com/28780306/51368532-cf2e3880-1b2a-11e9-808d-ed0fba4e75a4.png)

 I’m looking forward to your favourable reply!

**Comment by rahilwazir** [comment]
@felicityji No, I have not encountered segmentation fault error during any of the model training attempts.

**Comment by nhynes** [comment]
`mkdir results`

**Comment by rahilwazir** [comment]
@nhynes OK thanks. That worked and the pre-trained model ended successfully.

Now, what are my next steps? How can I execute the model? We have discussed this here already: https://github.com/torralba-lab/im2recipe-Pytorch/issues/13
There is a new comment by a user.

**Comment by nhynes** [comment]
>  what are my next steps?

Well it depends on what you wan to do. If it's to get results for your own dataset, just package everything up in the same format as the existing resources but use your own images and recipes. If you want to visualize the outputs, you'll want to load the embeddings in `img_embeds.pkl` and the like in tensorboard or your own plotting library.

**Comment by nhynes** [comment]
> Can you tell me how am I going to load the embeddings in img_embeds.pkl

I haven't seen the file format in a while, so I can't tell you what the schema is, but you can figure it out by exploration using a jupyter notebook:
```
import pickle
with open('img_embeds.pkl') as f_embs:
    embs = pickle.load(f_embs)
```
and then in the interactive cells, just
```
list(embs.keys())
list(embs['somekey'].keys())
len(embs['somekey']['embs'])
# etc
```

> Also, is that file easily importable in tensorboard?

Almost definitely not, but the converter won't be more than 100 lines of code. If you can send me the schema of the embeddings pickle file, I can help you write the converter.

> Sorry for asking so many questions, I'm new to this.

No problem! I'm truly glad that you're able to get something out of this project :)

**Comment by BornaGhotbi** [comment]
@felicityji I had the same problem running the code on Python 3.7.
Changing it to Python 3.6.8 :: Anaconda, Inc. worked for me.

**Comment by Asanebou** [comment]
@rahilwazir  Hi, I notice a problem in your screenshots above, which also occurs when I run the code. It seems in your screenshots the validation loss MedR doesn't decrease (it keeps a value around 500) even after more than 200 epochs. (However, the  pretrained model shows a medR of only 4.7 in epoch 220.) Have you got a better result when you finished the running?

**Comment by BornaGhotbi** [comment]
I am facing the same problem. Only my rec loss decreases during training. Cos loss and image loss are constant even after 100 epochs which leads to MedR of 500 for the results. @nhynes Do you know how should the training losses go down?

**Comment by felicityji** [comment]
@nhynes  Hi,I am facing the same problem when i run  python train.py.  I have found that many people reacted to this problem, and i very much hope that you can give the advise on this code. I am very interested in your research work. Thank  you so much.
![image](https://user-images.githubusercontent.com/28780306/54658502-a0234b80-4b08-11e9-8bc6-0a2dda6b2e36.png)
